src/components/verification-grid/verification-grid.ts

```diff
@@ -292,6 +338,194 @@ export class VerificationGridComponent extends AbstractComponent(LitElement) {
+        if (refinedTarget % i === 0) {
+          const columns = i;
+          const rows = refinedTarget / i;
+
+          const candidateShape = { columns, rows };
+          const willFit = this.willFitTileSize(containerSize, candidateShape);
+          const meetsAspectRatio = this.gridAspectRatioSimilarity(containerSize, candidateShape) > targetThreshold;
+
+          if (willFit) {
+            hasFittingCandidate = true;
+
+            if (meetsAspectRatio) {
+              candidateShapes.push({ columns, rows });
+            }
+          }
+        }
+      }
+
+      // if we have reached this condition, there is a fatal bug in the
+      // dynamic grid size code. This condition should never be triggered
+      // I have added the condition so that if dynamic grid size gets stuck in
+      // and infinite loop it won't completely crash the browser
+      // meaning that the verification grid will soft fail to prevent all the
+      // users work being lost
+      const maxGridSize = 1_000 as const;
+      if (refinedTarget >= maxGridSize || refinedTarget <= 0) {
+        console.error("Reached maximum/minimum grid size. Recovering via a 1x1 grid.");
+        this.targetGridSize = 1;
+        return { columns: 1, rows: 1 };
+      }
+
+      // there are two reasons why we might reject candidate shapes
+      // either they would not fit on the screen or they do not meet the
+      // minimum similarity to the containers aspect ratio
+      //
+      // if all the candidates are rejected because they do not fit, we will
+      // attempt to decrease the size until we find a target that correctly fits
+      //
+      // if we had a candidate that fit, but we rejected it because it did not
+      // meet the aspect ratio threshold, we continue to increase the target
+      // until we find a candidate that meets our conditions
+      if (candidateShapes.length === 0) {
+        if (!hasFittingCandidate) {
+          if (isPrime(refinedTarget)) {
+            refinedTarget += 1;
+            continue;
+          }
+
+          if (refinedTarget === 1) {
+            return { columns: 1, rows: 1 };
+          }
+
+          refinedTarget -= 1;
+          continue;
+        }
+
+        refinedTarget += 1;
+        continue;
+      }
+
+      // to find the optimal grid shape, we find what grid shape will result in
+      // the largest grid tile size
+      // this also inadvertently finds the grid shape that will best match the
+      // containers aspect ratio
+      let optimalGridShape: GridShape = candidateShapes[0];
+      for (const candidate of candidateShapes) {
+        const optimalSize = this.tileSize(containerSize, optimalGridShape);
+        const proposedSize = this.tileSize(containerSize, candidate);
+
+        const optimalArea = optimalSize.width * optimalSize.height;
+        const proposedArea = proposedSize.width * proposedSize.height;
+
+        if (proposedArea > optimalArea) {
```

Comment:
naming alone: how can a proposal be greater than optimal?

Reply:
This loop finds the most optimal candidate, meaning that we check if the `propsedArea > foundOptimalArea`

In hindsight: 

We should find the optimal once and store the `optimalArea` value instead of re-computing it every iteration